Move the SQS input to use the V2 version of the AWS library

ph opened the pull request:
This PR includes the following changes:
- Uses V2 version of the library, which resolve the memory leak issue.
- Replace our custom poller with the `QueuePoller` from the AWS library
- Replaced `run_with_backoff` with a non recursive function.
- Adding unit test for the handling of events.
- Adding a small integration test that push an event to a real queue and fetch it.

This PR is build upon @dwagner2301's PR https://github.com/logstash-plugins/logstash-input-sqs/pull/16

Fixes: #17 #11

dwagner2301 committed:
AWS ruby SDK v2 upgrade

API-Calls changed to support the AWS ruby SDK v2. Unsupported config
parameters and exceptions removed.

dwagner2301 committed:
Replaces aws-sdk dependencies with mixin-aws

To depend on aws sdk is not necessary. To use logstash-mixin-aws as a
dependency is much better solution.

dwagner2301 committed:
Removes unnecessary de-allocation

The de-allocation is not neccessary since the GC should take care
of it as soon as the plugin isn't used anymore

ph committed:
Move the code into smaller methods to allow easier mocking and testing

jsvd commented on lib/logstash/inputs/sqs.rb:
```diff
@@ -76,101 +77,108 @@ class LogStash::Inputs::SQS < LogStash::Inputs::Threadable
+                   end
+  end
+
+  def verify_credentials!
```
I find it strange that the `verify_credentials!` has the side effect of populating `@queue_url` (and also `@sqs` through `aws_sqs_client`)

maybe make `register` do a call to a `setup_queue` method that, as a consequence, can report a credential failure?

ph replied:
Completely agree!

jsvd commented on lib/logstash/inputs/sqs.rb:
```diff
@@ -76,101 +77,108 @@ class LogStash::Inputs::SQS < LogStash::Inputs::Threadable
+  def polling_options
+    { 
+      :max_number_of_messages => MAX_MESSAGES_TO_FETCH,
+      :attribute_names => ["SentTimestamp"],
```
if `max_number_of_messages` is a constant, then `attribute_names` should be too, I think

ph replied:
+1

jsvd commented on lib/logstash/inputs/sqs.rb:
```diff
@@ -76,101 +77,108 @@ class LogStash::Inputs::SQS < LogStash::Inputs::Threadable
+
+  def handle_message(message)
+    event = decode_event(message)
+    decorate(event)
```
I think I'd expect that if I use

```
sqs {
  ..
  ..
  add_field => ["test", "test"]
}
```

Field `test` will always have `test` in it when leaving the plugin, but if you specify `id_field` to be "test" then it will be replaced since `decorate` happens before `add_sqs_data`.

Thoughts?

ph replied:
I think your idea makes a lot of sense and I'll change the code to reflect that.
But I think the `decorate` function should happen after a plugin execution so maybe we can add some code to the pipeline that will automatically call `decorate` on newly created events?

ph replied:
We could also add metadata to tell people that we have replaced a field? (thinking out loud)

jsvd commented on lib/logstash/inputs/sqs.rb:
```diff
@@ -76,101 +77,108 @@ class LogStash::Inputs::SQS < LogStash::Inputs::Threadable
+        retry
+      end # retry limit exceeded
+    rescue LogStash::ShutdownSignal
+    rescue Exception => e
```
Rescuing exception is quite harsh, maybe StandardError or RuntimeError is enough?

ph replied:
agree

jsvd commented on lib/logstash/inputs/sqs.rb:
```diff
@@ -76,101 +77,108 @@ class LogStash::Inputs::SQS < LogStash::Inputs::Threadable
+        next_sleep = sleep_time * BACKOFF_FACTOR 
+        retry
+      end # retry limit exceeded
+    rescue LogStash::ShutdownSignal
```
I'd add a simple comment here saying something like "do nothing, gracefully return" or similar

ph replied:
+1

jsvd commented on lib/logstash/inputs/sqs.rb:
```diff
@@ -76,101 +77,108 @@ class LogStash::Inputs::SQS < LogStash::Inputs::Threadable
-  public
-  def run(output_queue)
-    @logger.debug("Polling SQS queue", :queue => @queue)
+  def queue_url
```
changing the assignment 1 line below allows better indentation:

``` ruby
def queue_url
  @queue_url ||= aws_sqs_client.get_queue_url(:queue_name =>  @queue)[:queue_url]
rescue Aws::SQS::Errors::ServiceError => e
  @logger.error("Cannot establish connection to Amazon SQS", :error => e)
  raise LogStash::ConfigurationError, "Verify the SQS queue name and your credentials"
end
```

jsvd commented on lib/logstash/inputs/sqs.rb:
```diff
@@ -76,101 +77,108 @@ class LogStash::Inputs::SQS < LogStash::Inputs::Threadable
   # Name of the event field in which to store the  SQS message Sent Timestamp
```
"the  SQS" -> "the SQS"

ph replied:
Ah double space! :D

ph commented:
@jsvd I have addressed your comments in the latest commits.
Also fixed a few more things:
- the spec files do not require an internet connection to run
- the integration test also check the the credentials.

jsvd commented on lib/logstash/inputs/sqs.rb:
```diff
@@ -73,104 +74,105 @@ class LogStash::Inputs::SQS < LogStash::Inputs::Threadable
-  end # def run_with_backoff
+      next_sleep = sleep_time
+    rescue Aws::SQS::Errors::ServiceError => e
+      @logger.info("Aws::SQS::Errors::ServiceError ... retrying SQS request with exponential backoff", :queue => @queue, :sleep_time => sleep_time, :error => e)
```
how often does this occur? maybe I'd bump it to warn or error, users need to be aware that the there's a weak link in the chain

ph replied:
agree.

jsvd commented on lib/logstash/inputs/sqs.rb:
```diff
@@ -73,104 +74,105 @@ class LogStash::Inputs::SQS < LogStash::Inputs::Threadable
+    rescue Aws::SQS::Errors::ServiceError => e
+      @logger.info("Aws::SQS::Errors::ServiceError ... retrying SQS request with exponential backoff", :queue => @queue, :sleep_time => sleep_time, :error => e)
+      sleep(next_sleep)
+      next_sleep =  next_sleep > max_time ? sleep_time : next_sleep = sleep_time * BACKOFF_FACTOR 
```
why not just
`next_sleep =  next_sleep > max_time ? sleep_time : sleep_time * BACKOFF_FACTOR`

ph replied:
@jsvd Can you check if you agree with this logic? 

In the current version of the plugin we have a max retries of SQS specific errors before silently killing the input thread.
I've changed that so the backoff will keep retrying (using a max sleep time), I mean theses errors should be transient but 60 seconds could happen and the input should not be killed for that.

ph replied:
(aah typo for your previous comment.)

ph committed:
Add the option to configure polling frequency

ph commented:
@jsvd has found an interesting feature/bug.

When you try to shutdown the pipeline it will send  a `ShutdownSignal` using `Thread#raise` (which is bad) and the aws sdk is catching this error and use it as a retryable error. This behavior make logstash never stop gracefully.

ph commented:
Form what I see the `seahorse` http lib shouldnt not mark this exception as a retryable and should raise it back, still digging.

ph commented:
I have added a monkey patch to fix the issue and created an upstream issue in aws repo.
Added this https://github.com/aws/aws-sdk-ruby/issues/912

ph committed:
Adding a monkey patch to make sure `LogStash::ShutdownSignal` doesn't
get catch by AWS RetryError.

ph commented:
@jsvd It should fix the issue with the input when you are trying to shutdown, please do a final review so we can release it.

jsvd commented on spec/inputs/sqs_spec.rb:
```diff
@@ -1 +1,170 @@
+
+      context "when the option isn't specified" do
+        it "doesnt add the `message_id`" do
+          expect(subject[id_field]).to be_nil
```
maybe `expect(subject).to_not include(id_field)` ?

jsvd commented on spec/integration/sqs_spec.rb:
```diff
@@ -0,0 +1,111 @@
+      end
+
+      it "add the `message_id`" do
+        expect(subject[id_field]).not_to be_nil
```
similar comment as above: `expect(subject).to include(id_field)`

ph replied:
I always forgot about `include`, oh well :dancers:

jsvd commented:
Tested failing scenario, proper shutdown now! :+1: :+1: 
Tests run, code reviewed. Minor comments otherwise LGTM

ph committed:
update changelog

ph commented:
@jsvd I think it is worthy to make it a 2.0.0 release what do you think?

jsvd commented:
Well the api hasn't changed, only augmented, and for the end user the big change is that it works better than before so I'd suggest a 1.1.0
